File: code/2300.successful-pairs-of-spells-and-potions.php
<?php
/*
 * @lc app=leetcode id=2300 lang=php
 *
 * [2300] Successful Pairs of Spells and Potions
 */

// @lc code=start
class Solution {

    /**
     * @param Integer[] $spells
     * @param Integer[] $potions
     * @param Integer $success
     * @return Integer[]
     */
    function successfulPairs($spells, $potions, $success) {
        sort($potions);
        $result = [];
        for ($i = 0; $i < count($spells); $i++) {
            $count = 0;

            $low = 0;
            $high = count($potions) - 1;
            while ($low <= $high) {
                $mid = floor(($low + $high) / 2);

                if ($potions[$mid] * $spells[$i] < $success) {
                    $low = $mid + 1;
                } else {
                    $count += $high - $mid + 1;
                    $high = $mid - 1;
                }
            }

            $result[$i] = $count;
        }

        return $result;
    }
}
// @lc code=end

File: code/2300.successful-pairs-of-spells-and-potions.go
package code

import "sort"

/*
 * @lc app=leetcode id=2300 lang=golang
 *
 * [2300] Successful Pairs of Spells and Potions
 */

// @lc code=start
func successfulPairs(spells []int, potions []int, success int64) []int {
	ans := make([]int, len(spells))

	sort.Slice(potions, func(i, j int) bool {
		return potions[i] < potions[j]
	})

	for i, spell := range spells {
		minIndex := 0
		maxIndex := len(potions) - 1

		if potions[len(potions)-1]*spell < int(success) {
			ans[i] = 0
			continue
		}

		for true {
			cur := (minIndex + maxIndex) / 2

			if potions[cur]*spell >= int(success) {
				maxIndex = cur
			} else {
				minIndex = cur + 1
			}

			if maxIndex == minIndex {
				ans[i] = len(potions[minIndex:])
				break
			}
		}
	}

	return ans
}

// @lc code=end
